src/bigint-encoder.js

```diff
@@ -0,0 +1,97 @@
+/**
+ * Encode BigInt value from arbitrary provided arguments
+ *
+ * @param {Array<Number|BigInt|String>} parts - Slices to encode, where the
+ *    earlier elements represent higher bits in the final integer
+ * @param {64|128|256} size - Number of bits in the target integer type
+ * @param {Boolean} unsigned - Whether it's an unsigned integer
+ *
+ * @returns {BigInt}
+ */
+export function encodeBigIntFromBits(parts, size, unsigned) {
+  let result = 0n;
+  // check arguments length
+  if (parts.length && parts[0] instanceof Array) {
+    parts = parts[0];
+  }
+  const total = parts.length;
+  if (total === 1) {
+    try {
+      result = BigInt(parts[0]);
+      if (!unsigned) {
+        result = BigInt.asIntN(size, result);
+      }
+    } catch (e) {
+      throw new TypeError(`Invalid integer value: ${parts[0]}`)
+    }
+  } else {
+    const sliceSize = size / total;
+    if (sliceSize !== 32 && sliceSize !== 64 && sliceSize !== 128)
+      throw new TypeError('Invalid number of arguments')
+    // combine parts
+    for (let i = 0; i < total; i++) {
+      let part = BigInt.asUintN(sliceSize, BigInt(parts[i].valueOf()));
+      if (i > 0) { // shift if needed
+        part <<= BigInt(i * sliceSize);
+      }
+      result |= part;
```

Comment:
Actually, at the cost of some efficiency, I think you could maybe replace this entire if/else block (starting at `const total = parts.length;`) with:
```js
const sliceSize = size / parts.length;
if (sliceSize !== 32 && sliceSize !== 64 && sliceSize !== 128)
      throw new TypeError('Invalid number of arguments');
const shift = BigInt(sliceSize);
const result = parts
  .map(BigInt)
  .reduce((value, slice) => (value << shift) | BigInt.asUintN(sliceSize, slice), 0n);
if (!unsigned) { // clamp value to the requested size
  result = BigInt.asIntN(size, result);
}
```
Whether that's worth it or not idk.

But, I *think* regardless you could combine the `parts.length === 1` and `parts.length > 1` cases into a single flow.

Reply:
this is the kinda dank thing I was tryna do in a test case here: [scint_test.js#L295-L302](https://github.com/stellar/js-stellar-base/blob/ab0b4af99de004bbc7ee1f0a3352bdbc7f83bbc0/test/unit/scint_test.js#L295-L302).

I gave it a shot here: ba17a3a, but I'm not confident in it and there's a failing test:

```
  1) UnsignedHyper.fromString
       fails for negative numbers:
     AssertionError: expected [Function] to throw an error
      at Context.<anonymous> (test/unit/unsigned-hyper_test.js:89:61)
```

but that test:

```js
  it('fails for negative numbers', function () {
    expect(() => UnsignedHyper.fromString('-1059')).to.throw(/range/);
  });
```

doesn't seem like it should be a failure case anyway: why can't you instantiate an unsigned value from a signed one? cc @orbitlens

Reply:
> why can't you instantiate an unsigned value from a signed one?

@Shaptic 
Theoretically, you can. But what's the meaning of that? I we are talking about serializing bytes<->int that's a valid behavior. But if you are playing with real representations of the number (in a for of a `string` or `bigint`), this has no sense. Especially in terms of idempotent conversion (e.g. `UnsignedHyper.fromString('-1059').toString() !== '-1059'` -- the value is malformed).

Reply:
Hmm sure, good point!

I think I'm missing how the code handled this prior to me changing the loop :confused: I added an explicit check in ea4efc8, but it seems like it was done automatically previously and that bugs me heh

Reply:
Ahhhh I got it, it was handled by the `if (total === 1)` special case:

```js
  if (total === 1) {
    try {
      result = BigInt(parts[0]);
      if (!unsigned) {
        result = BigInt.asIntN(size, result);
      }
    } catch (e) {
      throw new TypeError(`Invalid integer value: ${parts[0]}`)
    }
  }
```

since `parts[0]` wasn't reinterpreted via `asUintN(64, ...`, it won't be in the right range :100:

Reply:
@paulbellamy @Shaptic 
To tell the truth, I don't like the `map`+`reduce` solution. The original `for` loop looks much cleaner than then code I see in [ba17a3a](https://github.com/stellar/js-xdr/commit/ba17a3a179019ceccfe8c95651a088ee2ecdbff1) commit. Not to mention that it's slower and memory inefficient (arrow functions not add function call + context binding, while the return values are copied each time, because bigint is a value type, not a reference type).

Reply:
ah which means, previously, `encodeBigIntFromBits([-1, -1], size: 64, unsigned: true)` was valid, but `encodeBigIntFromBits([-1], size: 64, unsigned: true)` was not??

Reply:
> ah which means, previously, encodeBigIntFromBits([-1, -1], size: 64, unsigned: true) was valid, but encodeBigIntFromBits([-1], size: 64, unsigned: true) was not??

Yeah, that's counter-intuitive, agree with you. `encodeBigIntFromBits([-1], size: 64, unsigned: true)` should be valid, but `encodeBigIntFromBits(-1n, size: 64, unsigned: true)` should fail. At least as I understand it.

Reply:
Good points re: allocation in the `reduce`. I conflated trying to simplify the loop with trying to eliminate the `if/else` special case on `total` in the same commit; it would have been better to separate them.

Wdyt about it now (as of 6d3484c)? I moved the code back to using loops. It should still read easier than the previous code on account of no conditional, and should be faster due to eliding some allocations/conversions/iterations.

Reply:
@Shaptic Looks good 👍 
Thanks for iterating on this.

Reply:
thx for solid feedback :+1: